rustc suggests unstable library method on stable

### Code

```Rust
fn main() {
    let v = vec![1,2,3,4];
    for pair in v.chunk(2) {
        println!("{}, {}", pair[0], pair[1]);
    }
}
```


### Current output

```Shell
error[E0599]: no method named `chunk` found for struct `Vec<{integer}>` in the current scope
 --> src/main.rs:3:19
  |
3 |     for pair in v.chunk(2) {
  |                   ^^^^^ help: there is a method with a similar name: `as_chunks`

For more information about this error, try `rustc --explain E0599`.
```


### Desired output

```Shell
error[E0599]: no method named `chunk` found for struct `Vec<{integer}>` in the current scope
 --> src/main.rs:3:19
  |
3 |     for pair in v.chunk(2) {
  |                   ^^^^^ help: there is a method with a similar name: `chunks`

For more information about this error, try `rustc --explain E0599`.
```


### Rationale and extra context

It would be more helpful to suggest the `chunks` method for the typo rather than a nightly-only symbol which isn't actually available.

1. Following the suggestion doesn't work.
2. The `chunks()` method is actually available on a `std::slice`.
3. The `chunks()` method has been stable since 1.0.0.
4. I had trouble noticing the difference between `.chunk()` and `.chuncks()` and could have used the compiler's help.

### Other cases

_No response_

### Anything else?

Reproduced with 1.68.0 stable and 1.70 nightly.

<!-- TRIAGEBOT_START -->

<!-- TRIAGEBOT_ASSIGN_START -->

<!-- TRIAGEBOT_ASSIGN_DATA_START$${"user":"Ezrashaw"}$$TRIAGEBOT_ASSIGN_DATA_END -->

<!-- TRIAGEBOT_ASSIGN_END -->
<!-- TRIAGEBOT_END -->

--- a/compiler/rustc_hir_typeck/src/method/probe.rs
+++ b/compiler/rustc_hir_typeck/src/method/probe.rs
@@ -1027,6 +1027,15 @@ impl<'a, 'tcx> ProbeContext<'a, 'tcx> {
                     true
                 }
             })
+            // ensure that we don't suggest unstable methods
+            .filter(|candidate| {
+                // note that `DUMMY_SP` is ok here because it is only used for
+                // suggestions and macro stuff which isn't applicable here.
+                !matches!(
+                    self.tcx.eval_stability(candidate.item.def_id, None, DUMMY_SP, None),
+                    stability::EvalResult::Deny { .. }
+                )
+            })
             .map(|candidate| candidate.item.ident(self.tcx))
             .filter(|&name| set.insert(name))
             .collect();

--- a/tests/ui/stability-attribute/auxiliary/similar-unstable-method.rs
+++ b/tests/ui/stability-attribute/auxiliary/similar-unstable-method.rs
@@ -1,0 +1,13 @@
+#![feature(staged_api)]
+#![stable(feature = "libfoo", since = "1.0.0")]
+
+#[unstable(feature = "foo", reason = "...", issue = "none")]
+pub fn foo() {}
+
+#[stable(feature = "libfoo", since = "1.0.0")]
+pub struct Foo;
+
+impl Foo {
+    #[unstable(feature = "foo", reason = "...", issue = "none")]
+    pub fn foo(&self) {}
+}

--- a/tests/ui/stability-attribute/issue-109177.rs
+++ b/tests/ui/stability-attribute/issue-109177.rs
@@ -1,0 +1,13 @@
+// aux-build: similar-unstable-method.rs
+
+extern crate similar_unstable_method;
+
+fn main() {
+    // FIXME: this function should not suggest the `foo` function.
+    similar_unstable_method::foo1();
+    //~^ ERROR cannot find function `foo1` in crate `similar_unstable_method` [E0425]
+
+    let foo = similar_unstable_method::Foo;
+    foo.foo1();
+    //~^ ERROR no method named `foo1` found for struct `Foo` in the current scope [E0599]
+}

--- a/tests/ui/stability-attribute/issue-109177.stderr
+++ b/tests/ui/stability-attribute/issue-109177.stderr
@@ -1,0 +1,21 @@
+error[E0425]: cannot find function `foo1` in crate `similar_unstable_method`
+  --> $DIR/issue-109177.rs:7:30
+   |
+LL |     similar_unstable_method::foo1();
+   |                              ^^^^ help: a function with a similar name exists: `foo`
+   |
+  ::: $DIR/auxiliary/similar-unstable-method.rs:5:1
+   |
+LL | pub fn foo() {}
+   | ------------ similarly named function `foo` defined here
+
+error[E0599]: no method named `foo1` found for struct `Foo` in the current scope
+  --> $DIR/issue-109177.rs:11:9
+   |
+LL |     foo.foo1();
+   |         ^^^^ method not found in `Foo`
+
+error: aborting due to 2 previous errors
+
+Some errors have detailed explanations: E0425, E0599.
+For more information about an error, try `rustc --explain E0425`.

